magenta-lib/src/main/scala/magenta/deployment_type/GcpDeploymentManager.scala

```diff
@@ -0,0 +1,158 @@
+      |Number of seconds to wait for the deployment operations to complete
+      |""".stripMargin
+  ).default(1800) // half an hour
+
+  val deploymentNameParam: Param[String] = Param(
+    name = "deploymentName",
+    documentation =
+      """
+        |The name of the deployment that this config should be applied to
+        |""".stripMargin
+  )
+
+  val configPathParam: Param[String] = Param[String](
+    name = "configPath",
+    documentation =
+      """
+        |The path of the config in the package
+        |""".stripMargin,
+    optional = true // one of this or configPathByStage
+  )
+
+  val configPathByStageParam: Param[Map[String, String]] = Param[Map[String, String]](
+    name = "configPathByStage",
+    documentation =
+      """
+        |A dict of stages to paths of config files in the package. When the current stage is found in here the config file
+        |will be used from this dict. If it's not found here then it will fall back to `configPath` if it exists.
+        |```
+        |configPathByStage:
+        |  PROD: config-file-prod.yaml
+        |  CODE: config-file-code.yaml
+        |```
+        |""".stripMargin,
+    optional = true // one of this or configPath
+  )
+
+  val upsertParam: Param[Boolean] = Param[Boolean](
+    name = "upsert",
+    documentation =
+      """
+        |When true a deployment should be created if it doesn't already exist. When false a deployment will fail
+        |if it doesn't exist but still updated if it does exist.
+        |""".stripMargin
+  ).default(false)
+
+  val previewParam: Param[Boolean] = Param[Boolean](
+    name = "preview",
+    documentation =
+      """
+        |Pass through preview to any update or insert operation.
+        |""".stripMargin
+  ).default(false)
+
+  val updateAction: Action = Action(
+    name = "update",
+    documentation =
+      """
+        |Update the deployment
+        |""".stripMargin
+  )
+  { (pkg, resources, target) =>
+    {
+      implicit val keyRing: KeyRing = resources.assembleKeyring(target, pkg)
+      val reporter: DeployReporter = resources.reporter
+
+      val projectName: String = keyRing.apiCredentials.get("gcp").map(_.id).getOrElse{
+        reporter.fail(s"No GCP key on keyring so unable to find project name. Ensure there is a credentials:gcp entry in ${resources.lookup.name} for ${pkg.app}, ${target.parameters.stage}, ${target.stack}")
+      }
+      implicit val artifactClient: S3Client = resources.artifactClient
+
+      val maxWaitDuration = maxWaitParam(pkg, target, reporter).seconds
+      val deploymentName = deploymentNameParam(pkg, target, reporter)
+      val upsert = upsertParam(pkg, target, reporter)
+      val preview = previewParam(pkg, target, reporter)
+
+      val configPathLookup: PartialFunction[String, String] = (configPathByStageParam.get(pkg), configPathParam.get(pkg)) match {
+        case (Some(map), None) => map
+        case (None, Some(default)) => { case _ => default }
```

Comment:
cant this just be `=> default`?

Reply:
It's expecting a `PartialFunction[String, String]` and `default` is a String, hence the `{ case _ => default}`. Perhaps I can express this more clearly by omitting the intermidiate lookup function.

Reply:
I've cleaned it up to make it simpler. Improves the error messaging too.

Reply:
nice